File: packages/flask/korea_auth/providers/apple.py
"""
Apple provider implementation
"""

import requests
import jwt
from ..result import AuthResult


class AppleProvider:
    """Verify Apple login tokens."""
    
    APPLE_KEYS_URL = "https://appleid.apple.com/auth/keys"
    
    def verify(self, token: str) -> AuthResult:
        """Verify an Apple identity token (JWT) and return user info."""
        try:
            # Fetch Apple's public keys
            keys_response = requests.get(self.APPLE_KEYS_URL)
            if not keys_response.ok:
                return AuthResult.failure("Failed to fetch Apple public keys")
            
            jwks = keys_response.json()
            
            # Decode header to find matching key
            unverified_header = jwt.get_unverified_header(token)
            kid = unverified_header.get("kid")
            
            # Find the matching key
            key_data = None
            for key in jwks.get("keys", []):
                if key.get("kid") == kid:
                    key_data = key
                    break
            
            if not key_data:
                return AuthResult.failure("No matching key found")
            
            # Build public key and verify
            public_key = jwt.algorithms.RSAAlgorithm.from_jwk(key_data)
            payload = jwt.decode(
                token,
                public_key,
                algorithms=["RS256"],
                options={"verify_aud": False}  # You should verify aud in production
            )
            
            return AuthResult.success_result(
                uid=payload["sub"],
                provider="apple",
                name=None,  # Apple doesn't always provide name in the token
                email=payload.get("email"),
                photo_url=None,
                raw_data=payload,
            )
        except jwt.InvalidTokenError as e:
            return AuthResult.failure(f"JWT error: {str(e)}")
        except Exception as e:
            return AuthResult.failure(str(e))

File: packages/rails/lib/korea_auth/providers/apple.rb
# frozen_string_literal: true

require "jwt"
require "httparty"

module KoreaAuth
  module Providers
    class Apple < Base
      APPLE_KEYS_URL = "https://appleid.apple.com/auth/keys"

      def verify(token)
        # Decode and verify Apple's identity token (JWT)
        # Fetch Apple's public keys
        keys_response = HTTParty.get(APPLE_KEYS_URL)
        
        unless keys_response.success?
          return Result.failure("Failed to fetch Apple public keys")
        end

        jwks = keys_response.parsed_response

        # Decode the JWT header to find the right key
        header = JWT.decode(token, nil, false).last
        key_data = jwks["keys"].find { |k| k["kid"] == header["kid"] }

        unless key_data
          return Result.failure("No matching key found")
        end

        # Build the public key
        jwk = JWT::JWK.new(key_data)
        
        # Decode and verify the token
        decoded = JWT.decode(token, jwk.public_key, true, { algorithm: "RS256" })
        payload = decoded.first

        Result.success(
          uid: payload["sub"],
          provider: "apple",
          name: nil, # Apple doesn't always provide name in the token
          email: payload["email"],
          photo_url: nil,
          raw_data: payload
        )
      rescue JWT::DecodeError => e
        Result.failure("JWT decode error: #{e.message}")
      rescue StandardError => e
        Result.failure(e.message)
      end
    end
  end
end
